storage/src/adb/immutable.rs

```diff
@@ -302,14 +311,41 @@ impl<E: RStorage + Clock + Metrics, K: Array, V: Codec, H: CHasher, T: Translato
         assert_eq!(log_size, leaf_pos_to_num(mmr.size()).unwrap());
         assert_eq!(log_size, locations.size().await?);
 
-        Ok(log_size)
+        Ok((log_size, oldest_retained_loc.unwrap_or(0)))
     }
 
     /// Returns the section of the log where we are currently writing new items.
     pub fn current_section(&self) -> u64 {
         self.log_size / self.log_items_per_section
     }
 
+    /// Return the oldest location that remains retrievable.
+    pub fn oldest_retained_loc(&self) -> Option<u64> {
+        if self.log_size == 0 {
+            None
+        } else {
+            Some(self.oldest_retained_loc)
+        }
+    }
+
+    /// Prunes the db of up to all operations that have location less than `loc`. The actual number
+    /// pruned may be fewer than requested due to blob boundaries.
+    pub async fn prune(&mut self, loc: u64) -> Result<(), Error> {
+        assert!(loc <= self.log_size);
+
+        // Prune the log up to the section containing the requested pruning location.
+        let section = loc / self.log_items_per_section;
+        self.log.prune(section).await?;
+        self.oldest_retained_loc = section * self.log_items_per_section;
+
+        // Prune the MMR & locations map up to the oldest retained item in the log after pruning.
+        self.locations.prune(self.oldest_retained_loc).await?;
```

Comment:
I generally think it makes sense to prune the "query path" (mmr/locations) before the "underlying" data (log) but don't think it should matter if we support prune alignment during init (see earlier comment).

Reply:
I use the log pruning boundary as the "source of truth" which means I need to prune it first to get the actual oldest_retained, and then the mmr/log boundary gets pruned to that location to ensure all retained log operations remain "covered".

Reply:
Let me try and better document this.

Reply:
Added documentation explaining why log is pruned first.